File: assignment1/c_impl/sortering.c
#include <stdio.h>

// QUICK SORT ALGORITHM


int partition(int arr[], const int p, const int r) {
    const int x = arr[r];
    int i = p - 1;
    int temp;
    for (int j = p; j < r; j++) {
        if (arr[j] <= x) {
            i++;
            temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    temp = arr[i + 1];
    arr[i + 1] = arr[r];
    arr[r] = temp;
    return i + 1;
}



void quick_sort(int arr[], const int p, const int r) {
    if (p < r) {
        const int q = partition(arr, p, r);
        quick_sort(arr, p, q - 1);
        quick_sort(arr, q + 1, r);
    }
}


// MERGE SORT ALGORITHM

void merge(int arr[], const int left, const int mid, const int right) {
    const int n_left = mid - left + 1;
    const int n_right = right - mid;

    int left_arr[n_left], right_arr[n_right];
    for (int i = 0; i < n_left; i++) {
        left_arr[i] = arr[left + i];
    }
    for (int i = 0; i < n_right; i++) {
        right_arr[i] = arr[mid + i +1];
    }

    int i = 0, j = 0, k = left;
    while (i < n_left && j < n_right) {
        if (left_arr[i] <= right_arr[j]) {
            arr[k++] = left_arr[i++];
        } else {
            arr[k++] = right_arr[j++];
        }
    }

    while (i < n_left)
        arr[k++] = left_arr[i++];

    while (j < n_right)
        arr[k++] = right_arr[j++];

}

void merge_sort(int arr[], const int left, const int right) {
    if (left >= right) {
        return;
    }

    const int mid = left + (right - left) / 2;
    merge_sort(arr, left, mid);
    merge_sort(arr, mid + 1, right);
    merge(arr, left, mid, right);
}


// INSERTION SORT

void insertion_sort(int arr[], const int n) {
    for (int i = 1; i < n; i++) {
        int j = i;
        int cur = arr[i];

        while (j > 0 && cur < arr[j-1]) {
            arr[j] = arr[j - 1];
            j--;
        }
        arr[j] = cur;
    }
}

// HEAP SORT

int left(const int i) {
    return  2 * i + 1;
}

int right(const int i) {
    return 2 * i + 2;
}

void max_heapify(int arr[], const int n, const int i) {
    int l = left(i);
    int r = right(i);
    int largest = i;

    if (l < n && arr[l] > arr[largest])
        largest = l;
    if (r < n && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        max_heapify(arr, n, largest);
    }
}

void build_max_heap(int arr[], const int n) {
    for (int i = n / 2 - 1; i >= 0; i--) {
        max_heapify(arr, n, i);
    }
}

void heap_sort(int arr[], const int n) {
    if (n <= 1)
        return;

    build_max_heap(arr, n);
    for (int i = n - 1; i >= 1; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        max_heapify(arr, i, 0);
    }
}



void print_array(int arr[], const int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}


int main(void) {
    int arr[] = {10, 7, 8, 9, 1, 5};
    constexpr int n = sizeof(arr) / sizeof(arr[0]);
    quick_sort(arr, 0, n - 1);
    printf("Sorted array: \n");
    print_array(arr, n);
    printf("\n");

    printf("Testing merge sort:\n");
    int arr2[] = {12, 11, 13, 5, 6, 7};
    constexpr int m = sizeof(arr2) / sizeof(arr2[0]);
    merge_sort(arr2, 0, m - 1);
    printf("Sorted array: \n");
    print_array(arr2, m);
    printf("\n");

    printf("Test insertion sort:\n");
    int arr3[] = {12, 11, 13, 5, 6};
    constexpr int k = sizeof(arr3) / sizeof(arr3[0]);
    insertion_sort(arr3, k);
    printf("Sorted array: \n");
    print_array(arr3, k);
    printf("\n");

    printf("Test heap sort:\n");
    int arr4[] = {12, 11, 13, 5, 6, 7};
    constexpr int l = sizeof(arr4) / sizeof(arr4[0]);
    heap_sort(arr4, l);
    printf("Sorted array: \n");
    print_array(arr4, l);
    return 0;
}

File: assignment1/python_impl/sortering.py

def insertion_sort(arr: list) -> int:
    steps = 0
    for i in range(1, len(arr)):
        j = i
        current = arr[i]
        steps += 1  # comparison in loop condition

        while j > 0 and current < arr[j-1]:
            steps += 2  # two comparisons in while condition
            arr[j] = arr[j-1]
            j -= 1
            steps += 1  # assignment operation

        arr[j] = current
        steps += 1  # final assignment

    return steps




def heap_sort(arr):
    steps = [0]

    def left(i):
        return 2*i+1

    def right(i):
        return 2*i+2

    def max_heapify(heap_size, i):
        l = left(i)
        r = right(i)
        largest = i
        steps[0] += 3  # three assignments
        
        if l < heap_size and arr[l] > arr[largest]:
            steps[0] += 2  # two comparisons
            largest = l
        else:
            steps[0] += 1  # one comparison (short-circuit)
            
        if r < heap_size and arr[r] > arr[largest]:
            steps[0] += 2  # two comparisons
            largest = r
        else:
            steps[0] += 1  # one comparison

        if largest != i:
            steps[0] += 1  # comparison
            arr[i], arr[largest] = arr[largest], arr[i]
            steps[0] += 1  # swap operation
            max_heapify(heap_size, largest)

    def build_max_heap():
        heap_size = len(arr)
        for i in range((heap_size // 2) - 1, -1, -1):
            max_heapify(heap_size, i)

    n = len(arr)
    if n < 2:
        return steps[0]

    build_max_heap()

    for end in range(n - 1, 0, -1):
        arr[0], arr[end] = arr[end], arr[0]
        steps[0] += 1  # swap
        max_heapify(end, 0)

    return steps[0]

def merge_sort(A, left, right, steps=None):
    is_initial_call = steps is None
    if steps is None:
        steps = [0]
    
    if left >= right:
        steps[0] += 1
        return steps[0] if is_initial_call else None

    steps[0] += 1
    mid = left + (right - left) // 2
    merge_sort(A, left, mid, steps)
    merge_sort(A, mid + 1, right, steps)
    merge(A, left, mid, right, steps)
    
    return steps[0] if is_initial_call else None


def merge(A, left, mid, right, steps=None):
    if steps is None:
        steps = [0]
        
    n_left = mid - left + 1
    n_right = right - mid

    L = [0] * n_left
    R = [0] * n_right
    
    for i in range(n_left):
        L[i] = A[left + i]
        steps[0] += 1

    for i in range(n_right):
        R[i] = A[mid + i + 1]
        steps[0] += 1

    i = j = 0
    k = left
    while i < n_left and j < n_right:
        steps[0] += 2  # two comparisons in while
        if L[i] <= R[j]:
            steps[0] += 1  # comparison
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1
        k += 1
        steps[0] += 1  # assignment

    while i < n_left:
        steps[0] += 1  # comparison
        A[k] = L[i]
        i += 1
        k += 1
        steps[0] += 1

    while j < n_right:
        steps[0] += 1  # comparison
        A[k] = R[j]
        j += 1
        k += 1
        steps[0] += 1
        


def quick_sort(arr, p, r, steps=None):
    is_initial_call = steps is None
    if steps is None:
        steps = [0]
    
    if p < r:
        steps[0] += 1
        q = partition(arr, p, r, steps)
        quick_sort(arr, p, q - 1, steps)
        quick_sort(arr, q + 1, r, steps)
    else:
        steps[0] += 1
    
    return steps[0] if is_initial_call else None


def partition(arr, p, r, steps=None):
    if steps is None:
        steps = [0]
        
    x = arr[r]
    i = p - 1
    steps[0] += 1
    
    for j in range(p, r):
        if arr[j] <= x:
            steps[0] += 1  # comparison
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
            steps[0] += 1  # swap
        else:
            steps[0] += 1  # comparison

    arr[i + 1], arr[r] = arr[r], arr[i + 1]
    steps[0] += 1  # swap
    return i + 1